### ISSUE TITLE
old-style fancy extra-deps (in packages section) are broken in new stack unless you specify 'extra-dep: true'

### ISSUE THREAD

**Issue by kadoban** [commercialhaskell/stack#3728]
Old-style extra-deps that are specified in ```packages``` used to work reliably (as far as I know) without specifying ```extra-dep: true``` in old versions of stack. They no longer seem to work at all (in 1.6.3)

The current behavior is arguably justifiable, but *changing* the behavior that everyone has been using for a long time isn't good. It seems like lots of previously-working stack projects are now broken.

See the output below. Note that the only difference between the two ```stack.yaml``` files is if ```extra-dep: true``` is specified or not (it works if it's specified, doesn't if it's not)

The testcase is on github if you want to play with it https://github.com/kadoban/test-super-proj

### Clone Testcase (and shows version)

```bash
~$ stack --version
Version 1.6.3, Git revision b27e629b8c4ce369e3b8273f04db193b060000db (5454 commits) x86_64 hpack-0.20.0
~$ mkdir -p test/super-proj && git clone https://github.com/kadoban/test-super-proj test/super-proj
[...]
~$ cd test/super-proj
```

### Broken Version

```bash
~/test/super-proj (works-in-stack163)$ git checkout borked-in-stack163 
Switched to branch 'borked-in-stack163'
Your branch is up-to-date with 'origin/borked-in-stack163'.
~/test/super-proj (borked-in-stack163)$ stack clean && stack build && stack exec -- super-proj

Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for super-proj-0.1.0.0:
    dep-two must match -any, but the stack configuration has no specified version
needed since super-proj is a build target.
[...]
```

### Extra-dep Specification in Broken Version

```yaml
packages:
- .
- location: ./dep-one
- location:
    git: https://github.com/kadoban/test-dep-two
    commit: fbff976ba58f7c31c6338193acebed2e930c7be1
```

### Working Version

```bash
~/test/super-proj (borked-in-stack163)$ git checkout works-in-stack163-and-old 
Switched to branch 'works-in-stack163-and-old'
Your branch is up-to-date with 'origin/works-in-stack163-and-old'.
~/test/super-proj (works-in-stack163-and-old)$ stack clean && stack build && stack exec -- super-proj
[...builds stuff successfully...]
Completed 2 action(s).
omg it's a string
omg it's another string
```

### Extra-dep Specification in Working Version

```yaml
packages:
- .
- location: ./dep-one
  extra-dep: true
- location:
    git: https://github.com/kadoban/test-dep-two
    commit: fbff976ba58f7c31c6338193acebed2e930c7be1
  extra-dep: true
```

Edit: I confirmed that both of the above versions work fine in stack version 1.5.1. (though the version without the ```extra-dep: true``` gets a warning, it doesn't say anything about it not working in the future, it just says something about the behavior of `stack test`

**Comment by kadoban** [comment]
I tried to bisect this to find where this actually changed, but I didn't have much luck.

As far as I can tell, it *might* have narrowed it down to a certain feature branch or two that branched from after v1.4.something (according to git describe). It seems that at some point there was an error message when you tried to use an extra-dep like this without actually setting extra-dep: true, so anything that actually gave that error message I couldn't use to bisect anymore.

They all seemed to report their version as v1.4.something from 'git describe', so I just started skipping anything that described itself as such.

The bisection log (listing the commits I had to skip) is at https://gist.github.com/kadoban/270687c04e610eb4da47a142b2b09964 if anyone can make sense of it. Might not even be a useful start to an investigation into this though, so feel free to ignore.

I plan to at least check each released version and see what the behavior is on this.

**Comment by kadoban** [comment]
Okay, so despite those commits with the explicit error I was finding during the bisection, there appears to be *no* released version of stack back to v1.1.0 (got bored after that one) that disallowed using an extra-dep without setting it `extra-dep: true`. Odd.

So the following are the results of building/running the version with old-style extra-deps *without* setting `extra-dep: true`.

Here's the full results (all ranges inclusive)
- v1.1.0 - v1.2.0: No warning, no error, builds fine
- v1.3.0 - v1.5.1: Warning about "No extra-dep setting found for package at URL [...]", but builds fine
- v1.6.0.20171022 - v1.6.3: Error about plan construction, doesn't explicitly complain about the extra-dep, the extra-dep is just ineffective. Won't build.

### Conclusion

**So in conclusion, in v1.6.0 or thereabouts, we started breaking old-style extra-deps that a lot of people have been using for a long time now. I know I've personally gotten at least two users who ran into this, so it's definitely out there in the wild.**


### Full Messages for completion

Here's the full warning message:

```
No extra-dep setting found for package at URL:

https://github.com/kadoban/test-dep-two

This is usually a mistake, external packages should typically
be treated as extra-deps to avoid spurious test case failures.
```

Note that it doesn't say anything about that being deprecated or about it stopping working sometime in the future, it just says something about how `stack test` works.

And the full error message from new stack:

```
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for super-proj-0.1.0.0:
    dep-two must match -any, but the stack configuration has no specified version
needed since super-proj is a build target.

Some potential ways to resolve this:

  * Set 'allow-newer: true' to ignore all version constraints and build anyway.

  * You may also want to try using the 'stack solver' command.

Plan construction failed.
```

**Comment by snoyberg** [comment]
This is an intentionally backwards-incompatible change, and there are no plans to revert it
